src/main/java/com/iridium/iridiumskyblock/listeners/PlayerPortalListener.java

```diff
@@ -2,41 +2,72 @@
 
 import com.iridium.iridiumcore.utils.StringUtils;
 import com.iridium.iridiumskyblock.IridiumSkyblock;
+import com.iridium.iridiumskyblock.PermissionType;
+import com.iridium.iridiumskyblock.database.Island;
+import com.iridium.iridiumskyblock.database.User;
 import com.iridium.iridiumskyblock.managers.IslandManager;
 import com.iridium.iridiumskyblock.utils.LocationUtils;
+import com.iridium.iridiumskyblock.utils.PlayerUtils;
 import org.bukkit.World;
 import org.bukkit.event.EventHandler;
 import org.bukkit.event.Listener;
 import org.bukkit.event.player.PlayerPortalEvent;
 import org.bukkit.event.player.PlayerTeleportEvent;
 
 import java.util.Objects;
+import java.util.Optional;
 
 public class PlayerPortalListener implements Listener {
 
     @EventHandler(ignoreCancelled = true)
     public void onPlayerPortal(PlayerPortalEvent event) {
         IslandManager islandManager = IridiumSkyblock.getInstance().getIslandManager();
-        IridiumSkyblock.getInstance().getIslandManager().getIslandViaLocation(event.getFrom()).ifPresent(island -> {
-            if (event.getCause().equals(PlayerTeleportEvent.TeleportCause.NETHER_PORTAL)) {
-                if (IridiumSkyblock.getInstance().getConfiguration().netherIslands) {
-                    World world = Objects.equals(event.getFrom().getWorld(), islandManager.getNetherWorld()) ? islandManager.getWorld() : islandManager.getNetherWorld();
-                    event.setTo(island.getCenter(world));
+
+        final Optional<Island> island = IridiumSkyblock.getInstance().getIslandManager().getIslandViaLocation(event.getFrom());
+        if (!island.isPresent())
+            return;
+
+        final User user = IridiumSkyblock.getInstance().getUserManager().getUser(event.getPlayer());
+        if (!IridiumSkyblock.getInstance().getIslandManager().getIslandPermission(island.get(), user, PermissionType.PORTAL)) {
+            event.getPlayer().sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().cannotUsePortal.replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix)));
+            event.setCancelled(true);
+
+            // Teleport them back to the island to prevent constant chat spam.
+            if (event.getCause() == PlayerTeleportEvent.TeleportCause.END_PORTAL) {
+
+                if (island.get().isVisitable()) {
```

Comment:
what if its a private island but they are a member of the island? I should defo add an island.canVisit method that checks for island members, island trusts, island bans and island private :thinking: